src/Processors/QueryPlan/Optimizations/filterPushDown.cpp

```diff
@@ -208,34 +215,43 @@ size_t tryPushDownFilter(QueryPlan::Node * parent_node, QueryPlan::Nodes & nodes
+        std::vector<ASTTableJoin::Kind> kinds = {ASTTableJoin::Kind::Left};
+
+        /// For not full sorting merge join push down is for left table only, because left and right streams are not independent.
+        /// Only inner and left(/right) join are supported. Other types may generate default values for left table keys.
         /// So, if we push down a condition like `key != 0`, not all rows may be filtered.
-        if (table_join.kind() == ASTTableJoin::Kind::Inner || table_join.kind() == ASTTableJoin::Kind::Left)
+        if (join->allowPushDownToRight())
+            kinds.emplace_back(ASTTableJoin::Kind::Right);
+
+        for (const auto kind : kinds)
         {
-            const auto & left_header = join->getInputStreams().front().header;
-            const auto & res_header = join->getOutputStream().header;
-            Names allowed_keys;
-            const auto & source_columns = left_header.getNames();
-            for (const auto & name : source_columns)
+            if (table_join.kind() == ASTTableJoin::Kind::Inner || table_join.kind() == kind)
             {
-                /// Skip key if it is renamed.
-                /// I don't know if it is possible. Just in case.
-                if (!left_header.has(name) || !res_header.has(name))
-                    continue;
-
-                /// Skip if type is changed. Push down expression expect equal types.
-                if (!left_header.getByName(name).type->equals(*res_header.getByName(name).type))
-                    continue;
-
-                allowed_keys.push_back(name);
+                const auto & streams = join->getInputStreams();
+                const auto & input_header = kind == ASTTableJoin::Kind::Left ? streams.front().header : streams.back().header;
+                const auto & res_header = join->getOutputStream().header;
+                Names allowed_keys;
+                const auto & source_columns = input_header.getNames();
+                for (const auto & name : source_columns)
+                {
+                    /// Skip key if it is renamed.
+                    /// I don't know if it is possible. Just in case.
+                    if (!input_header.has(name) || !res_header.has(name))
+                        continue;
+
+                    /// Skip if type is changed. Push down expression expect equal types.
+                    if (!input_header.getByName(name).type->equals(*res_header.getByName(name).type))
+                        continue;
+
+                    allowed_keys.push_back(name);
+                }
+
+                const bool can_remove_filter
+                    = std::find(source_columns.begin(), source_columns.end(), filter->getFilterColumnName()) == source_columns.end();
+                if (auto updated_steps = tryAddNewFilterStep(parent_node, nodes, allowed_keys, can_remove_filter, kind == ASTTableJoin::Kind::Left ? 0 : 1))
+                    return updated_steps;
             }
```

Comment:
I think it would be more readable if you put this into a lambda which accepts `kind` as arguemnt instead of creating a vector and iterating it.